File: python code/mergeSort.py
import random
import time
import psutil

# Function to merge two sorted subarrays
def merge(arr, left, mid, right):
    start_index_left = left
    start_index_right = mid + 1
    temp_element = left
    temp = [0] * len(arr)

    while start_index_left <= mid and start_index_right <= right:
        if arr[start_index_left] < arr[start_index_right]:
            temp[temp_element] = arr[start_index_left]
            start_index_left += 1
        else:
            temp[temp_element] = arr[start_index_right]
            start_index_right += 1
        temp_element += 1

    while start_index_left <= mid:
        temp[temp_element] = arr[start_index_left]
        temp_element += 1
        start_index_left += 1

    while start_index_right <= right:
        temp[temp_element] = arr[start_index_right]
        temp_element += 1
        start_index_right += 1

    for temp_element in range(left, right + 1):
        arr[temp_element] = temp[temp_element]

# Function to perform merge sort
def merge_sort(arr, left, right):
    if left < right:
        mid = (left + right) // 2
        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)
        merge(arr, left, mid, right)

if __name__ == "__main__":
  
    # Get random sample size for testing
    size = int(input("Enter a size: "))
    array = [random.randint(0, 99) for _ in range(size)]

    # Getting the start time, CPU time, and memory usage
    start_time = time.time()

    # Function call
    merge_sort(array, 0, size - 1)

    # Getting the stop time, CPU time, and memory usage
    stop_time = time.time()

    # changing the seconds to micoseconds
    duration = (stop_time - start_time) * 1e6  
  
    # Print the execution time, CPU time, and memory usage
    print("Merge sort time is:", duration, "microseconds")

    # Get cpu usuage
    cpu_usage = psutil.cpu_percent()
    print("CPU usage is:", cpu_usage, "%")

    # Get memory usage
    memory_usage = psutil.virtual_memory().used / 1024 
    print("Memory usage:", memory_usage, "KB")

    # Print the sorted list
    print("Sorted array:", array)

File: Java/merge_sort.java
class Mergesort {
    static long startTime = System.nanoTime();


    void merge(int array[], int o, int p, int r){

        int x = p - o + 1;
        int y = r - p;
        int L[] =  new int[x];
        int R[] = new int[y];

        for (int i = 0; i < x; ++i){
            L[i] = array[o + i];
        }
            

        for(int j = 0; j < y; j++){
            R[j] = array[p + 1 + j];

        }
        int i = 0;
        int j = 0;

        int k = o;
        while(i < x && j < y){
            if(L[i] <= R[j]){
                array[k] = L[i];
                i++;
            }
            else{
                array[k] = R[j];
                j++;

            }
            k++;
            
        }
        while (i < x){
            array[k] = L[i];
            i++;
            k++;

        }
        while (j < y){
            array[k] = R[j];
            j++;
            k++;
        }

        

    }

    void sort(int array[], int z, int w){

        if (z < w){

            int m = (z + w) / 2;

            sort(array, z, m);
            sort(array, m + 1, w);

            merge(array, z, m, w);

        }
    }

    static void printArray(int array[]){
        int n = array.length;

        for(int i = 0; i < n; ++i){
            System.out.print(array[i] + " ");


        }
        System.out.println();


    }

    public static void main(String args[]){
        int array[] = {3,5,9,1,6,3,4,2};
        System.out.println("Given Array");
        printArray(array);

        Mergesort t = new Mergesort();
        t.sort(array, 0,array.length - 1);
        
        printArray(array);

        long endTime = System.nanoTime();
        long duration = (endTime - startTime);
        System.out.println(duration);
        

    }

    
}
